nucypher/blockchain/eth/sol/source/contracts/WorkLock.sol

```diff
@@ -0,0 +1,147 @@
+        depositRate = _depositRate;
+        refundRate = _refundRate;
+        lockedPeriods = _lockedPeriods;
+    }
+
+    /**
+    * @notice Bid for tokens by transferring ETH
+    **/
+    function bid() public payable returns (uint256 newClaimedTokens) {
+        require(block.timestamp >= startBidDate && block.timestamp <= endBidDate,
+            "Bid is open during a certain period");
+        WorkInfo storage info = workInfo[msg.sender];
+        info.depositedETH = info.depositedETH.add(msg.value);
+        uint256 claimedTokens = info.depositedETH.mul(depositRate);
+        require(claimedTokens >= minAllowableLockedTokens && claimedTokens <= maxAllowableLockedTokens,
+            "Claimed tokens must be within the allowed limits");
+        newClaimedTokens = msg.value.mul(depositRate);
+        allClaimedTokens = allClaimedTokens.add(newClaimedTokens);
+        require(allClaimedTokens <= token.balanceOf(address(this)),
+            "Not enough tokens in the contract");
+        emit Bid(msg.sender, msg.value, newClaimedTokens);
+    }
+
+    /**
+    * @notice Claimed tokens will be deposited and locked as stake in the MinersEscrow contract.
+    **/
+    function claim() public returns (uint256 claimedTokens) {
+        require(block.timestamp >= endBidDate, "Claiming tokens allowed after bidding is over");
+        WorkInfo storage info = workInfo[msg.sender];
+        require(!info.claimed, "Tokens are already claimed");
+        info.claimed = true;
+        claimedTokens = info.depositedETH.mul(depositRate);
+        info.workDone = escrow.setWorkMeasurement(msg.sender, true);
+        token.approve(address(escrow), claimedTokens);
+        escrow.deposit(msg.sender, claimedTokens, lockedPeriods);
+        emit Claimed(msg.sender, claimedTokens);
+    }
+
+    /**
+    * @notice Refund ETH for the work done
+    **/
+    function refund() public returns (uint256 refundETH) {
+        WorkInfo storage info = workInfo[msg.sender];
+        require(info.claimed, "Tokens are not claimed");
+        require(info.depositedETH > 0, "Nothing deposited");
+        uint256 currentWork = escrow.getWorkDone(msg.sender);
+        uint256 workDone = currentWork.sub(info.workDone);
+        require(workDone > 0, "No work has been done.");
+        refundETH = workDone.div(refundRate);
+        if (refundETH > info.depositedETH) {
+            refundETH = info.depositedETH;
+        }
+        if (refundETH == info.depositedETH) {
+            escrow.setWorkMeasurement(msg.sender, false);
+        }
+        info.depositedETH = info.depositedETH.sub(refundETH);
+        workDone = refundETH.mul(refundRate);
+        info.workDone = info.workDone.add(workDone);
+        emit Refund(msg.sender, refundETH, workDone);
+        msg.sender.transfer(refundETH);
+    }
+
+    /**
+    * @notice Get remaining work to full refund
+    **/
+    function getRemainingWork(address _miner) public view returns (uint256) {
```

Comment:
@szotov is this function for stakers to check how much more work they need to do to get back the rest of their deposit? Does it serve some other purpose?

Reply:
Right now only for checking by staker or anyone else